### PR TITLE
Reset qnode after execution error

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

While investigating Issue #5442, I was getting really confused by why my tensorflow execution was giving a torch error. I realized that when the execution fails, we do not properly reset `QNode.interface` back to `"auto"`.  

**Description of the Change:**

Pulls a block of the code to a helper method so that we can wrap it in a `try-finally` block.

**Benefits:**

We can continue using a qnode after a failure has occured during execution.

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-60047]

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -291,6 +291,8 @@
 
 <h3>Bug fixes 🐛</h3>
 
+* The `QNode` now resets if an error occurs during execution.
```
Comment (astralcai) [inline doc/releases/changelog-dev.md:294]:
```suggestion
* The `QNode` now resets if an error occurs during execution.
  [(#5449)](https://github.com/PennyLaneAI/pennylane/pull/5449)
```

Comment (lillian542) [inline doc/releases/changelog-dev.md:294]:
```suggestion
* The `QNode` interface now resets if an error occurs during execution.
```
Missing PR link

**Review by lillian542 — APPROVED** [review]
Looks good to me!

**Review by astralcai — APPROVED** [review]
Looks good!

**Pull request merged**